### ISSUE TITLE
[BasicAA] Incorrect alias analysis causing miscompile in slp-vectorizer

### ISSUE THREAD

**Issue by wlei-llvm** [llvm/llvm-project#98978]
Hi! We recently hit a miscompile issue that slp-vectorizer turned two 8 byte(memory overlap) load/store into one 16 byte load/store and we further pinpointed to an incorrect alias analysis.  
Here is the reduced LLVM IR to repro: [reduced.ll.txt](https://github.com/user-attachments/files/16240607/reduced.ll.txt)
Cmd:
```
opt -S -passes=slp-vectorizer  < reduced.ll
```
Incorrect codegen result:  (notice the `load <2 x i64>/  store <2 x i64> `)
```
15:                                               ; preds = %15, %14
  %16 = phi ptr [ %0, %15 ], [ %13, %14 ]
  %17 = phi ptr [ null, %15 ], [ %8, %14 ]
  %18 = load <2 x i64>, ptr %16, align 1
  %19 = shufflevector <2 x i64> %18, <2 x i64> poison, <2 x i32> <i32 1, i32 0>
  store <2 x i64> %19, ptr %17, align 1
  br i1 false, label %15, label %.loopexit42.2
```
It's a very tricky case, here are some explanations and pictures to help demo the issue:

1) To be simple, let's start using the node`%3` as root, the node value is a alias check for the same value alias(<%3, %3>) (this is because it's under the [`MaybeCrossIteration`(BasicAliasAnalysis.cpp)](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Analysis/BasicAliasAnalysis.cpp#L1804-L1805) to check if it's no-alias across different iterations )
2) Before recursively checking the dependencies on def-use chain, it initializes the cache with NoAlias(IIUC, this is an assumption based NoAlias [BasicAliasAnalysis.cpp#L1691-L1693)](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Analysis/BasicAliasAnalysis.cpp#L1691-L1693)
![demo1](https://github.com/user-attachments/assets/2a209c97-a4e9-40df-917c-6c85261768be)
3) Keep checking the alias recursively, i,e. check `<%23, %23>`, --> `<%7, %7>` -->`<%3, 3>` until it finds a circle. Then it runs the code in [BasicAliasAnalysis.cpp#L1693-L1702](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Analysis/BasicAliasAnalysis.cpp#L1693-L1702), which fetches the initial assumption based value in the cache for %3. 

4) Then the recursion return back to %7. IIUC, Now %7 is done all the dependencies check, it run the [code](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Analysis/BasicAliasAnalysis.cpp#L1725) to update stats to `definitive`. Since it depends on an assumption based value(%3), it's pushed into the `AssumptionBasedResults`([code](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Analysis/BasicAliasAnalysis.cpp#L1736-L1738)) which IIUC later if the assumption is disproven, we will remove all the results based on this assumption. 

5)  Continue returning to %23 and process %23's right child %8. 
![demo2](https://github.com/user-attachments/assets/b08ca785-20ee-4c07-b053-35cf2ed1ed49)

6)  (Bug occurs!) when processing %8's child %7, as previously %7 is marked as `definitive`, %8 is not considered as a  assumption based([code](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Analysis/BasicAliasAnalysis.cpp#L1694)) and never pushed into `AssumptionBasedResult`, this is incorrect because actually %7 is assumption based(its dependence %3 is assumption based and have't been proven yet), and <%8, %8> is missing to be pushed into  `AssumptionBasedResult`.
  
  
![demo3](https://github.com/user-attachments/assets/6c1a7f85-df95-41fa-8b85-cd5ec5142a97)

7) Return and process %3's right child %99, say <%99> is not a NoAlias, now the previous %3's NoAlias assumption is disproven, all the `AssumptionBaseResult` based on %3 are disproven, i,e. `<%23, %23>`, `<%7,%7>` are removed from cache and needed to recompute, but as %8 is missing in `AssumptionBasedResult`, <%8, %8> is still in the cache. Later any query uses the  <%8, %8> is incorrect, in this case, it causes the incorrect <%16, %17> alias that leads to the miscomplile. 

In summary:
There seems to be a bug in this assumption-disproven mechanism: we incorrectly turned a no-alias assumption into a definitive no-alias result, then later alias result (for other use of the def with incorrect result) were produced incorrectly (which should be assumption based result instead of definitive result) too. Finally when assumptions got invalidated correctly due to other alias on dependencies, we failed to invalidate part of the chain due to assumption based result being prematurely turned into definitive result, escaping invalidation.

Therefore, I'm posting here to ask for help how to properly fix this, thanks in advance! 

cc @nikic @WenleiHe  @fhahn @hiraditya @jdoerfert  @preames

**Comment by WenleiHe** [comment]
@nikic @preames thoughts on this one? we hit this during llvm-15 -> llvm-17 upgrade causing a miscompile on zstd compression library, and the miscompile was bisected to https://reviews.llvm.org/D136175, but based on Lei's investigation, that particular patch may not be the root cause and we may have a lurking bug in basic AA.

**Comment by hiraditya** [comment]
cc: @alexey-bataev if you can share your analysis, that'd be great!

**Comment by alexey-bataev** [comment]
> cc: @alexey-bataev if you can share your analysis, that'd be great!

Cannot add anything specific here, SLP relies on aliasing analysis to prove that it is safe to vectorize the code. Looks like here AA says that store i64 %19, ptr %17, align 1, and following load %21 do not alias, so it decides it is safe to vectorize.

**Comment by WenleiHe** [comment]
> > cc: @alexey-bataev if you can share your analysis, that'd be great!
> 
> Cannot add anything specific here, SLP relies on aliasing analysis to prove that it is safe to vectorize the code. Looks like here AA says that store i64 %19, ptr %17, align 1, and following load %21 do not alias, so it decides it is safe to vectorize.

Right SLP isn't doing anything wrong. The bug is in basic AA.

**Comment by nikic** [comment]
I think we'll need to add an additional state between "assumption" and "definitive". https://github.com/llvm/llvm-project/commit/c25fdb46e1ebc6ca03b9f3887d835ecff9f37ed3 is an incomplete sketch that fixes the test case (but misses handling to convert SemiDefinitive -> Definitive later and some other things).

**Comment by nikic** [comment]
I've put up https://github.com/llvm/llvm-project/pull/100130 with a candidate fix.
